src/nl/hannahsten/texifyidea/inspections/latex/probablebugs/LatexQuoteInspection.kt

```diff
@@ -0,0 +1,118 @@
+package nl.hannahsten.texifyidea.inspections.latex.probablebugs
+
+import com.intellij.codeInspection.InspectionManager
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiTreeUtil.findChildrenOfType
+import nl.hannahsten.texifyidea.inspections.InsightGroup
+import nl.hannahsten.texifyidea.inspections.TexifyInspectionBase
+import nl.hannahsten.texifyidea.inspections.TexifyRegexInspection
+import nl.hannahsten.texifyidea.psi.LatexNormalText
+import nl.hannahsten.texifyidea.util.files.commandsInFile
+import nl.hannahsten.texifyidea.util.magic.GeneralMagic
+import nl.hannahsten.texifyidea.util.magic.PatternMagic
+import nl.hannahsten.texifyidea.util.magic.RegexPattern
+import nl.hannahsten.texifyidea.util.requiredParameter
+import nl.hannahsten.texifyidea.util.toTextRange
+import java.util.regex.Pattern
+import javax.annotation.RegEx
+
+/**
+ * @author Michael Milton
+ */
+class LatexQuoteInspection : TexifyInspectionBase() {
+    override val inspectionGroup = InsightGroup.LATEX;
+    override val inspectionId = "AsciiQuotes";
+
+    /**
+     * Matches anything that looks like a LaTeX quote
+     */
+    private val quotePattern = """["'`]+""".toRegex();
+
+    /**
+     * Defines how to extract a quote pair from a command
+     */
+    data class QuoteMaker(val command: String, val openParam: Int, val closeParam: Int)
+
+    /**
+     * All the ways that csquotes defines a new quotation mark
+     */
+    private val quoteMakers = arrayOf(
+        QuoteMaker("\\MakeOuterQuote", 0, 0),
+        QuoteMaker("\\MakeInnerQuote", 0, 0),
+        QuoteMaker("\\MakeAutoQuote", 0, 1),
+        QuoteMaker("\\MakeForeignQuote", 1, 2),
+        QuoteMaker("\\MakeBlockQuote", 0, 2),
+        QuoteMaker("\\MakeForeignBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHyphenBlockQuote", 1, 3),
+        QuoteMaker("\\MakeHybridBlockQuote", 1, 3),
+    );
+
+    override fun inspectFile(file: PsiFile, manager: InspectionManager, isOntheFly: Boolean): List<ProblemDescriptor> {
+        // Build up a list of acceptable quote pairs. The ith index in both arrays must contain a matching quote
+        // pair. Additional quote pairs are extracted from the csquotes quote making commands
+        val openers = LinkedHashSet(arrayListOf("`", "``"));
+        val closers = LinkedHashSet(arrayListOf("'", "''"));
```

Comment:
Use Kotlin `listOf`

Reply:
For whatever reason the `LinkedHashSet` only has a constructor defined for a mutable list.

Reply:
But why use LinkedHashSet? This is how I would write it:

```kotlin
        val quotePairs = listOf(Pair("`", "'"), Pair("``", "''"))
        val openQuotes = quotePairs.map { it.first }
        val closeQuotes = quotePairs.map { it.second }
```

Reply:
Well because a `LinkedHashSet` is both ordered (which I need to ensure `openQuotes[i]` is paired with `closeQuotes[i]`), and also has O(1) lookups, which I want because I'm checking if `token in openQuotes` all the time.

Reply:
I have an idea: please use a list with O(n) lookup, and then run a profiler over TeXiFy (e.g. visualvm, see the wiki) over a full inspection run (e.g. after typing one letter in a document) until all inspections are finished, and then consider again what optimizations you are proposing and how much they help us in solving the performance problems.

Reply:
Look if you really prefer simplicity over efficiency I can remove this (setting up a profiler would be time consuming), but this is the kind `COMP101` basic optimisation I would always do, like using a dictionary instead of a list if I need to look up items by key. I don't think the increase in complexity (replacing one iterable class with another one with the same signature?) is even a consideration